### PR TITLE
Add E2E tests

### LINKED ISSUE #35: Add E2E tests
Add a suite of E2E tests to test the integration between the layer and the agent.

### PR DESCRIPTION
**Pull request by infiniteregrets**
This PR closes #35 adding tests conforming to the suite provided in #50 

Current state of the tests:
validation tests:
- run the node application with mirrord exec in a forked process
- send GET, POST, PUT, and DELETE requests and keep a counter
- POST validates data.txt, PUT creates a file with the data received, DELETE reads the file created by PUT, validates the data and removes the file. 
- if all operations succeed, we exit with EXIT_SUCCESS
- since only a single job and pod is created, we verify that these resources exist and are removed after agent timeout finishes. 

namespace tests [agent]:
- create a new namespace 
- run a binary with agent namespace assigned the new namespace
- check if the job and the pod exist in the created namespace within the timeout limit 
- run a binary with agent namespace assigned a non-existent namespace 
- make sure the process crashes
- verify no resources have been created in either the default or the bad namespace 

namespace tests [pod]:
- create a new namespace 
- run a binary with pod namespace assigned the new namespace in the form of an env var
- verify the job exists

### CODE REVIEW

**Comment by aviramha** [comment]
Great job! Now get it passing ;)

**Review by eyalb181 — CHANGES_REQUESTED** [review]
So we're going to need a pretty major rewrite here. Here are my main critiques:

- We want the test code to be easy to understand, so bash isn't a great choice. Lets go with Rust, since we're already using it and it's best not to introduce new languages without sufficient reason.
- Lets use a testing framework, not just write freeform code. It'll give our tests structure, would print out legible output, let us e.g. skip certain tests in the future, etc.
- Strive to minimize tests to check one thing. Instead of checking all four methods in one test, write a test for each. We want our test results to focus us as much as possible on where the problem is.
- Strive to have as much of the testing logic in the test code itself, rather than in the app being tested.

File: e2e/src/tests.rs
Comment (aviramha) [inline e2e/src/tests.rs:105]: I don't like this function.. It's flow is very odd and weird. I suggest it'll get the stdout reader, server and service url
and then it would do something like
```rust
request(service_url, "GET")
assert_eq!(stdout.read.., "GET: Request completed"
```

**Review by eyalb181 — CHANGES_REQUESTED** [review]

**Comment by infiniteregrets** [comment]
Todo:
- remove CURL requests, switch to reqwest 
- change CLI commands to kube-rs fn calls
- change how validate_data function works
- add more assertions/panics

File: e2e/src/tests.rs
Comment (aviramha) [inline e2e/src/tests.rs:42]: I'm not sure why you check that server is running here, we shouldn't care AFAIK

File: e2e/src/tests.rs
Comment (aviramha) [inline e2e/src/tests.rs:183]: Why would we need a sleep at the end of the test?

File: e2e/src/tests.rs
Comment (aviramha) [inline e2e/src/tests.rs:199]: just call wait here with timeout instead of sleeping

File: e2e/src/utils.rs
Comment (aviramha) [inline e2e/src/utils.rs:101]: why not just pass the method as enum instead of bytes?

File: e2e/src/tests.rs
Comment (infiniteregrets) [reply e2e/src/tests.rs:42]: for example if the server errors out for some reason on start, it will write to it's stderr and on line 44 we are reading the stdout and it will block forever, we dont exactly know what could happen so we need to check if all went i.e. server is running or it errored out. its just a guess that the server will be up and running in 5 secs (worst case)

File: e2e/src/tests.rs
Comment (infiniteregrets) [reply e2e/src/tests.rs:183]: some of the tests are using the same namespace, if I dont sleep then if one test deleted the namespace and the other wants to create it immediately and somehow it wasnt deleted then we get an error from the server. however in this case its a unique namespace so we can probably remove it. ill check and update

File: e2e/src/tests.rs
Comment (infiniteregrets) [reply e2e/src/tests.rs:199]: there is no method for a timeout in the impl for the Child process struct?

File: e2e/src/tests.rs
Comment (infiniteregrets) [reply e2e/src/tests.rs]: fixed, we don't need the sleep there

File: e2e/src/utils.rs
Comment (infiniteregrets) [reply e2e/src/utils.rs:101]: done

File: e2e/src/tests.rs
Comment (aviramha) [reply e2e/src/tests.rs:199]: https://docs.rs/tokio/latest/tokio/time/fn.timeout.html

File: e2e/src/tests.rs
Comment (aviramha) [reply e2e/src/tests.rs:43]: you should read the stdout when the process finishes, there's no value in reading it in real time.
flow should be:
1. start process
2. send requests, do stuff
3. kill/verify process is done
4. validate stderr/stdout of process

**Review by eyalb181 — CHANGES_REQUESTED** [review]

File: e2e/src/tests.rs
Comment (infiniteregrets) [reply e2e/src/tests.rs:21]: thanks! i completely ignored that. fixed

File: e2e/src/tests.rs
Comment (infiniteregrets) [reply e2e/src/tests.rs:71]: takes around ~35 to cleanup the job & pod, i will change the sleep. if you think it should be different then please let me know!

File: e2e/src/tests.rs
Comment (infiniteregrets) [reply e2e/src/tests.rs:28]: yes, changed it to `HashMap::new()`

File: e2e/scripts/install.sh
Comment (infiniteregrets) [reply e2e/scripts/install.sh:6]: thanks for pointing out, fixed now

File: .github/workflows/ci.yaml
Comment (infiniteregrets) [reply .github/workflows/ci.yaml:88]: didn't realize we could add this manually through the action, fixed now

File: e2e/src/utils.rs
Comment (aviramha) [inline e2e/src/utils.rs:28]: relative path, use https://docs.rs/bintest/latest/bintest/

File: e2e/src/tests.rs
Comment (aviramha) [inline e2e/src/tests.rs:1]: move tests to tests dir

Comment (aviramha) [reply e2e/src/tests.rs:1]: and rename file to something like sanity

File: .github/workflows/ci.yaml
Comment (eyalb181) [inline .github/workflows/ci.yaml:72]: Can we try and do a platform matrix here? Lets start with mac and ubuntu.

File: tests/src/sanity.rs
Comment (eyalb181) [inline tests/src/sanity.rs:80]: Why are you asserting this?

File: tests/src/sanity.rs
Comment (eyalb181) [inline tests/src/sanity.rs:49]: I'd call this validation_timeout, not validation_duration. The validation duration should (hopefully) be a lot shorter.

File: tests/src/sanity.rs
Comment (eyalb181) [inline tests/src/sanity.rs:81]: As discussed (I think):
We want to try and not sleep the whole 35 seconds, if possible. So lets do something like this:

counter = 0
job_len = 1
while counter =< 35 && job_len > 0
{
job_len = pods_api...
sleep(1)
counter++
}
assert(job_len, 0)

File: tests/src/sanity.rs
Comment (eyalb181) [inline tests/src/sanity.rs:88]: Maybe verify the name here too, to make the code clearer and more robust

File: tests/src/sanity.rs
Comment (eyalb181) [inline tests/src/sanity.rs:101]: Why does it panic without the sleep?

File: tests/src/sanity.rs
Comment (eyalb181) [inline tests/src/sanity.rs:97]: Lets extract all this setup code to an external function instead of repeating it in each test

File: tests/src/sanity.rs
Comment (eyalb181) [inline tests/src/sanity.rs:31]: Why is this necessary?

File: .github/workflows/ci.yaml
Comment (eyalb181) [reply .github/workflows/ci.yaml:72]: Save this for later though - let's get this PR merged first.

File: tests/src/sanity.rs
Comment (eyalb181) [inline tests/src/sanity.rs:132]: Lets try to keep all the implicit knowledge of what the API does in one place. You already have validate_requests, so maybe add validate_no_requests next to it, or something like that.

Comment (eyalb181) [reply tests/src/sanity.rs:132]: Just to elaborate - right now, if I make a change to how app.js work, I would have to go over the entire code and look for dependencies. So instead we should strive to cenralize all code that relies on the internal implementation of app.js in one place.

File: tests/src/utils.rs
Comment (eyalb181) [inline tests/src/utils.rs:53]: Rather than querying all the pods and filtering them in memory, you can filter them in the api call and then just call .first(). See pod_api.rs in mirrord_layer for an example.

Comment (eyalb181) [reply tests/src/utils.rs:53]: This can wait for after the merge.

File: tests/src/utils.rs
Comment (eyalb181) [inline tests/src/utils.rs:102]: A. try a clearer function name
B. as a rule, lets not have asserts in util functions, only in the tests themselves

File: tests/src/utils.rs
Comment (eyalb181) [inline tests/src/utils.rs:122]: Isn't this the same as app.yaml? Why not use it?

**Review by eyalb181 — CHANGES_REQUESTED** [review]

File: tests/src/sanity.rs
Comment (infiniteregrets) [reply tests/src/sanity.rs:80]: so the pod name contains the name of the job, just making sure i selected the correct job

File: Cargo.toml
Comment (infiniteregrets) [reply Cargo.toml:10]: i am doing this because one of the tests, i believe test_agent_image fails and iirc it was something related to manifest not found. I can rerun the tests with tests being a part of the workspace members and ping you if that error happens again

File: .github/workflows/ci.yaml
Comment (infiniteregrets) [reply .github/workflows/ci.yaml:72]: yes, i'll add macos

File: tests/src/sanity.rs
Comment (infiniteregrets) [reply tests/src/sanity.rs:31]: we cannot panic from inside the thread. even if we panic the test doesnt fail so we need to change how the default panic works

File: .github/workflows/ci.yaml
Comment (infiniteregrets) [reply .github/workflows/ci.yaml:72]: added to my todo list, will open an issue later

File: tests/src/sanity.rs
Comment (infiniteregrets) [reply tests/src/sanity.rs:49]: reduced the duration to 20 seconds

File: tests/src/sanity.rs
Comment (infiniteregrets) [inline tests/src/sanity.rs:83]: @aviramha @meowjesty do you guys think this timeout on job cleanup is okay? or is there a better way to do it?

**Review by meowjesty — COMMENTED** [review]
Feedback regarding sleeps and watcher.

**Pull request closed without merging**
